NATIVE_POSTGRES_JSON_COMPARED_TO_EQL.md

```diff
@@ -0,0 +1,510 @@
+SELECT plaintext_jsonb->0 FROM examples;
+
+-- `->>` (returns text)
+SELECT plaintext_jsonb->>0 FROM examples;
+```
+
+### EQL
+
+EQL JSONB functions accept an eJSONPath as an argument (instead of using `->`/`->>`) for lookups.
+
+#### Decryption example
+
+EQL currently doesn't support returning a specific array element for decryption, but `cs_ste_vec_value_v1` can be used to return an array to the client to process.
+
+The query...
+
+```sql
+SELECT cs_ste_vec_value_v1(encrypted_jsonb, $1) AS val FROM examples;
+```
+
+With the params...
+
+```javascript
+// Assume that examples.encrypted_jsonb has JSON objects with the shape:
+{
+  "field_a": [1, 2, 3]
+}
+
+// `$1` is the EQL plaintext payload for the eJSONPath `$.field_a`:
+{
+  "k": "pt",
+  "p": "$.field_a",
+  "i": {
+    "t": "examples",
+    "c": "encrypted_jsonb"
+  },
+  "v": 1,
+  "q": "ejson_path"
+}
+```
+
+Would return the EQL plaintext payload with an array (`[1, 2, 3]` for example):
+
+```javascript
+// Example result for a single row
+{
+  "k": "pt",
+  "p": "[1, 2, 3]",
+  "i": {
+    "t": "examples",
+    "c": "encrypted_jsonb"
+  },
+  "v": 1,
+  "q": null
+}
+```
+
+#### Comparison example
+
+`cs_ste_vec_terms_v1` can be used with the native Postgres array access operator to get a term for comparison by array index.
+
+The eJSONPath used with `cs_ste_vec_terms_v1` needs to end with `[*]` (`$.some_array_field[*]` for example).
+
+> [!IMPORTANT]
+> Array access with `cs_ste_vec_terms_v1` only works when the given eJSONPath only matches a single array.
+> Accessing array elements from `cs_ste_vec_terms_v1` when the eJSONPath matches multiple arrays (for example, when there are nested arrays or multiple arrays at the same depth) can return unexpected results.
+
+The following query compares the first item in the array at the eJSONPath in $1 to the value in $2.
+
+```sql
+SELECT * FROM examples
+WHERE (cs_ste_vec_terms_v1(examples.encrypted_jsonb, $1))[1] > cs_ste_vec_term_v1($2)
+```
+
+```javascript
+// Assume that examples.encrypted_jsonb has JSON objects with the shape:
+{
+  "field_a": [4, 5, 6]
+}
+
+// `$1` is the EQL plaintext payload for the eJSONPath `$.field_a[*]`:
+{
+  "k": "pt",
+  "p": "$.field_a[*]",
+  "i": {
+    "t": "examples",
+    "c": "encrypted_jsonb"
+  },
+  "v": 1,
+  "q": "ejson_path"
+}
+
+// `$2` is the EQL plaintext payload for the ORE term to compare against (in this case, the ORE term for the integer `3`):
+{
+  "k": "pt",
+  "p": "3",
+  "i": {
+    "t": "examples",
+    "c": "encrypted_jsonb"
+  },
+  "v": 1,
+  "q": "ste_vec"
+}
+```
+
+## `json #>> text[]` → `text` and `json #> text[]` → `jsonb`/`json`
+
+### Native Postgres JSON(B)
+
+```sql
+-- `#>` (returns JSON(B))
+SELECT plaintext_jsonb#>'{field_a,field_b}' FROM examples;
+
+-- `#>>` (returns text)
+SELECT plaintext_jsonb#>>'{field_a,field_b}' FROM examples;
+```
+
+### EQL
+
+EQL JSONB functions accept an eJSONPath as an argument (instead of using `#>`/`#>>`) for lookups.
```

Comment:
Same again — define or link eJSONPath.

Reply:
Fixed in https://github.com/cipherstash/encrypt-query-language/pull/61/commits/beabf2bd22ddd7489755bcb60e25aaaadc48bc20 and https://github.com/cipherstash/encrypt-query-language/pull/61/commits/9224ad68fcc389a792f253397b66f6287f2205c8.